### DISCUSSION TITLE
Generic effect and interpose

### DISCUSSION THREAD

**Discussion by mkleczek** [haskell-effectful/effectful#150] [Q&A] [answered]
A newbie question. How to handle generic effect like here:
```
data MyEff a :: Effect where
  MyEff :: a -> MyEff a m ()

type instance DispatchOf (MyEff a) = 'Dynamic

tracedSql :: (MyEff a :> es) => Eff es () -> Eff es ()
tracedSql = interpose $ \env (MyEff a) -> do
    send $ MyEff a

```

This does not compile with a message: Could not deduce (MyEff a0 :> es) arising from a use of ‘send’
Am I missing something obvious here?

**Comment by ocharles** [comment]
I believe this is what effectful-plugin is for

**Reply by mkleczek** [reply]
I see. It should be possible to add some explicit type annotations and have it compiled without the plugin, no?

**Reply by arybczak** [reply, marked as answer]
Yes, this works:

```haskell
data MyEff a :: Effect where
  MyEff :: a -> MyEff a m ()
type instance DispatchOf (MyEff a) = Dynamic

tracedSql :: forall a es. (MyEff a :> es) => Eff es () -> Eff es ()
tracedSql = interpose @(MyEff a) $ \env (MyEff a) -> do
    send $ MyEff a
```

Plugin doesn't really help if there is no type annotation on `interpose` in this case for some reason, ~~not sure exactly why~~.

EDIT: It doesn't help because the plugin disambiguates type variables of a particular effect, not effects themselves.

**Reply by mkleczek** [reply]
Thanks, works brilliantly
